connectors/byoc.py

```diff
@@ -181,96 +181,116 @@ async def all_connectors(self):
-        self.job_id = resp["_id"]
-        return self.job_id
+    @property
+    def configuration(self):
+        return DataSourceConfiguration(self.get("connector", "configuration"))
+
+    @property
+    def filtering(self):
+        return Filter(self.get("connector", "filtering", default={}))
 
-    async def done(self, indexed_count=0, deleted_count=0, exception=None):
-        self.completed_at = datetime.now(timezone.utc)
+    @property
+    def pipeline(self):
+        return Pipeline(self.get("connector", "pipeline"))
 
-        job_def = {
-            "deleted_document_count": deleted_count,
-            "indexed_document_count": indexed_count,
-            "completed_at": iso_utc(self.completed_at),
-        }
+    @property
+    def terminated(self):
+        return self.status in (JobStatus.ERROR, JobStatus.COMPLETED, JobStatus.CANCELED)
 
-        if exception is None:
-            self.status = JobStatus.COMPLETED
-            job_def["error"] = None
-        else:
-            self.status = JobStatus.ERROR
-            job_def["error"] = str(exception)
+    @property
+    def indexed_document_count(self):
+        return self.get("indexed_document_count", default=0)
 
-        job_def["status"] = self.status.value
+    @property
+    def indexed_document_volume(self):
+        return self.get("indexed_document_volume", default=0)
 
-        return await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    @property
+    def deleted_document_count(self):
+        return self.get("deleted_document_count", default=0)
 
-    async def suspend(self):
-        self.status = JobStatus.SUSPENDED
-        job_def = {"status": self.status.value}
+    @property
+    def total_document_count(self):
+        return self.get("total_document_count", default=0)
 
-        await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    async def validate_filtering(self, validator):
+        validation_result = await validator.validate_filtering(self.filtering)
 
-    @classmethod
-    def transform_filtering(cls, filtering):
-        # deepcopy to not change the reference resulting in changing .elastic-connectors filtering
-        filtering = (
-            {"advanced_snippet": {}, "rules": []}
-            if (filtering is None or len(filtering) == 0)
-            else deepcopy(filtering)
+        if validation_result.state != FilteringValidationState.VALID:
+            raise InvalidFilteringError(
+                f"Filtering in state {validation_result.state}, errors: {validation_result.errors}."
+            )
+
+    async def claim(self):
+        doc = {
+            "status": JobStatus.IN_PROGRESS.value,
+            "started_at": iso_utc(),
+            "last_seen": iso_utc(),
+            "worker_hostname": socket.gethostname(),
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
+
+    async def done(self, ingestion_stats={}, connector_metadata={}):
```

Comment:
You can't have this signature:
```
async def done(self, ingestion_stats={}, connector_metadata={}):
```

do:

```
async def done(self, ingestion_stats=None, connector_metadata=None):
    if ingestion_stats is None:
        ingestion_stats = {}
```

reason: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Reply:
TIL

Reply:
Done in https://github.com/elastic/connectors-python/pull/460/commits/25427141df45140874ae859b947218f2c950d726